alicloud/resource_alicloud_snapshot.go

```diff
@@ -0,0 +1,149 @@
+package alicloud
+
+import (
+	"fmt"
+	"github.com/aliyun/alibaba-cloud-sdk-go/services/ecs"
+	"github.com/hashicorp/terraform/helper/resource"
+	"github.com/hashicorp/terraform/helper/schema"
+	"github.com/terraform-providers/terraform-provider-alicloud/alicloud/connectivity"
+	"time"
+)
+
+func resourceAliyunSnapshot() *schema.Resource {
+	return &schema.Resource{
+		Create: resourceAliyunSnapshotCreate,
+		Read:   resourceAliyunSnapshotRead,
+		Update: resourceAliyunSnapshotUpdate,
+		Delete: resourceAliyunSnapshotDelete,
+
+		Importer: &schema.ResourceImporter{
+			State: schema.ImportStatePassthrough,
+		},
+
+		Schema: map[string]*schema.Schema{
+			"disk_id": &schema.Schema{
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+			},
+			"name": &schema.Schema{
+				Type:     schema.TypeString,
+				Optional: true,
+				ForceNew: true,
+			},
+			"description": &schema.Schema{
+				Type:     schema.TypeString,
+				Optional: true,
+				ForceNew: true,
+			},
+			"tags": tagsSchema(),
+		},
+	}
+}
+
+func resourceAliyunSnapshotCreate(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*connectivity.AliyunClient)
+
+	args := ecs.CreateCreateSnapshotRequest()
+	args.DiskId = d.Get("disk_id").(string)
+	if name, ok := d.GetOk("name"); ok {
+		args.SnapshotName = name.(string)
+	}
+	if description, ok := d.GetOk("description"); ok {
+		args.Description = description.(string)
+	}
+
+	raw, err := client.WithEcsClient(func(ecsClient *ecs.Client) (interface{}, error) {
+		return ecsClient.CreateSnapshot(args)
+	})
+	if err != nil {
+		return fmt.Errorf("create snapshot failed, %s", err)
+	}
+	resp := raw.(*ecs.CreateSnapshotResponse)
+	if resp == nil {
+		return fmt.Errorf("create snapshot failed, the response is nil")
+	}
+	d.SetId(resp.SnapshotId)
+
+	ecsService := EcsService{client}
+	if err := ecsService.WaitUntilSnapshotCreatingCompeleted(d.Id()); err != nil {
+		return fmt.Errorf("wait for snapshot (%s) creating completed failed, %s", d.Id(), err)
```

Comment:
Line 69: I think the method `WaitUntilSnapshotCreatingCompeleted` should be a common method which contains id, status and timeout. This common method can be re-used in the delete method.